manager/allocator/network/port/port_test.go

```diff
@@ -0,0 +1,955 @@
+package port_test
+
+import (
+	. "github.com/docker/swarmkit/manager/allocator/network/port"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"fmt"
+
+	"github.com/docker/swarmkit/api"
+)
+
+var _ = Describe("port.Allocator", func() {
+	var (
+		pa *Allocator
+	)
+
+	BeforeEach(func() {
+		pa = NewAllocator()
+	})
+
+	Describe("adding a new endpoint", func() {
+		Context("to a clean port allocator", func() {
+			var (
+				spec, specCopy *api.EndpointSpec
+				endpoint       *api.Endpoint
+				p              Proposal
+				err            error
+			)
+			BeforeEach(func() {
+				spec = &api.EndpointSpec{
+					Ports: []*api.PortConfig{
+						{
+							Name:          "http",
+							Protocol:      api.ProtocolTCP,
+							TargetPort:    80,
+							PublishedPort: 80,
+							PublishMode:   api.PublishModeIngress,
+						},
+						{
+							Name:          "https",
+							Protocol:      api.ProtocolTCP,
+							TargetPort:    443,
+							PublishedPort: 443,
+							PublishMode:   api.PublishModeIngress,
+						},
+					},
+				}
+				endpoint = &api.Endpoint{}
+				specCopy = spec.Copy()
+				p, err = pa.Allocate(endpoint, spec)
+			})
+			It("should succeed", func() {
+				Expect(err).NotTo(HaveOccurred())
+				Expect(p).NotTo(BeNil())
+			})
+			It("should not modify the spec", func() {
+				Expect(spec).To(Equal(specCopy))
+			})
+			It("should not modify the endpoint", func() {
+				e := &api.Endpoint{}
+				Expect(endpoint.Spec).To(Equal(e.Spec))
+				Expect(endpoint.Ports).To(BeEmpty())
+			})
+			It("should not modify the Allocator state before calling Commit", func() {
+				// calling Allocate a second time will succeed because no state
+				// was committed
+				_, err := pa.Allocate(&api.Endpoint{}, spec)
+				Expect(err).ToNot(HaveOccurred())
+			})
+			It("should be idempotent", func() {
+				// grab a copy of the endpoint so we have something concrete to
+				// compare to
+				endpointCopy := endpoint.Copy()
+				endpointCopy.Spec = spec
+				endpointCopy.Ports = p.Ports()
+				p.Commit()
```

Comment:
Non-blocking:  would it make sense to add an extra check here before the commit to assert it's not a noop?

Reply:
Sure, this test can actually probably be reworked a bit to verify that allocator state changes AFTER commit as well, but I thought I had that test someplace else...

Reply:
Added a check for not noop to the "should succeed" case.